PluginStructs/P248_data_struct.cpp

```diff
@@ -0,0 +1,262 @@
+#ifdef USES_P248
+
+#include "../PluginStructs/P248_data_struct.h"
+
+
+uint8_t AHTSetCalCmd[3]    = { AHTX_CMD, 0x08, 0x00 }; // load factory calibration coeff
+uint8_t AHTSetCycleCmd[3]  = { AHTX_CMD, 0x28, 0x00 }; // enable cycle mode
+uint8_t AHTMeasureCmd[3]   = { 0xAC, 0x33, 0x00 };     // start measurement command
+uint8_t AHTResetCmd        =   0xBA;                   // soft reset command
+
+
+P248_data_struct::P248_data_struct(uint8_t addr) :
+  last_hum_val(0.0f),
+  last_temp_val(0.0f),
+  last_measurement(0),
+  i2cAddress(addr),
+  state(AHT_state::Uninitialized) {}
+
+
+bool P248_data_struct::initialized() const {
+  return state != Uninitialized;
+}
+
+void P248_data_struct::setUninitialized() {
+  state = Uninitialized;
+}
+
+// Only perform the measurements with big interval to prevent the sensor from warming up.
+bool P248_data_struct::update(unsigned long task_index) {
+  const unsigned long current_time = millis();
+
+  if (!initialized()) {
+    if (!begin()) {
+      return false;
+    }
+    state            = Initialized;
+    last_measurement = 0;
+  }
+
+  if (state != Wait_for_samples) {
+    if ((last_measurement != 0) &&
+        !timeOutReached(last_measurement + (Settings.TaskDeviceTimer[task_index] * 1000))) {
```

Comment:
I'm not entirely sure this is the best approach, for the plugin to look at the TaskDeviceTimer.
What if the command `taskRun` is called? Then it won't collect a sample.
So maybe better to let the scheduler handle this and then just call a re-schedule as soon as you set the state for starting a measurement.
You can however still use this TaskDeviceTimer value to re-schedule the task to get back in sync with the original interval by keeping track of the moment the PLUGIN_READ was called when it wasn't waiting for a sample.

Reply:
ok